### PR TITLE
zh_keyboard: add pinyin input-method

### LINKED ISSUE #9833: FR: [Add Chinese pinyin input method（添加中文拼音输入法）]
**Does your feature request involve difficulty completing a task? Please describe.**
Add Chinese pinyin input method,I think the Stroke count method is too few, at least I've never used it and can't master it.

添加中文拼音输入法，使用笔画输入法的人太少了，至少我没用过，我尝试了并发现很难掌握。（这能实现吗，期待中国人的回复）

### PR DESCRIPTION
**Pull request by weijiuqiao**
Simple pinyin ime, usable but not smart. See #9833
Closes #9833.

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/9843)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]

**Comment by greatyingzi** [comment]
牛B呀，o(￣▽￣)ｄ。

**Comment by greatyingzi** [comment]
我上周想做这个拼音输入法来着，刚搞完词库，停住了，没想到你就弄出来了。之前没写过lua，先改下插件练练手，想着以后做一个类似“阅读”的插件，可以配置源的。哈哈哈

File: frontend/ui/widget/virtualkeyboard.lua [hunk 1]
```diff
@@ -799,11 +799,13 @@ local VirtualKeyboard = FocusManager:extend{
         tr = "tr_keyboard",
         vi = "vi_keyboard",
         zh = "zh_keyboard",
+        zh_CN = "zh_CN_keyboard",
```
Comment (poire-z) [inline frontend/ui/widget/virtualkeyboard.lua:802]: I don't want to complexify that, and have another mapping for keyboard names, but just asking: will this be clear/guessable enough, that the first is stoke-based and the second is pinyin ?
https://github.com/koreader/koreader/blob/61364ffc3341f4ee7ff3de0816c4a8e176a4e470/frontend/ui/language.lua#L51-L52

File: frontend/ui/data/keyboardlayouts/zh_CN_keyboard.lua [hunk 2]
```diff
@@ -0,0 +1,155 @@
+local IME = require("frontend/ui/data/keyboardlayouts/generic_ime")
+local util = require("util")
+local _ = require("gettext")
+
+-- Start with the english keyboard layout (deep copy, to not alter it)
+local py_keyboard = require("util").tableDeepCopy(require("ui/data/keyboardlayouts/en_keyboard"))
+local SHOW_CANDI_KEY = "keyboard_chinese_pinyin_show_candidates"
```
Comment (poire-z) [inline frontend/ui/data/keyboardlayouts/zh_CN_keyboard.lua:7]: SETTING_NAME instead of KEY - to make it more obvious what it is.

**Review by poire-z — APPROVED** [review]

File: frontend/ui/data/keyboardlayouts/generic_ime.lua [hunk 3]
```diff
@@ -364,6 +368,24 @@ function IME:wrappedAddChars(inputbox, char, orig_char)
...
+            elseif imex.index <= self.last_index + 1 then
+                return
+            else
+                imex.index = imex.index - 1
+                imex.char = imex.candi[imex.index - self.last_index]
+            end
+        elseif #imex.candi > 1 then
+            imex.index = math.max(imex.index-1, 1)
+            local remainder = imex.index % #imex.candi
+            imex.char = imex.candi[remainder==0 and #imex.candi or remainder]
```
Comment (greatyingzi) [inline frontend/ui/data/keyboardlayouts/generic_ime.lua:384]: "=="两边添加空格好一些

Comment (poire-z) [reply frontend/ui/data/keyboardlayouts/generic_ime.lua:384]: אני מרגיש מודר

File: frontend/ui/widget/virtualkeyboard.lua (hunk 1 above)
Comment (weijiuqiao) [reply frontend/ui/widget/virtualkeyboard.lua:802]: >will this be clear/guessable enough, that the first is stoke-based and the second is pinyin

Not really. It says the second is most likely pinyin, so the first has to be something else ;)

File: frontend/ui/data/keyboardlayouts/generic_ime.lua (hunk 3 above)
Comment (weijiuqiao) [reply frontend/ui/data/keyboardlayouts/generic_ime.lua:384]: As it's within the brackets, tighter would be better :)
>אני מרגיש מודר

C'est drôle :)

File: frontend/ui/data/keyboardlayouts/zh_CN_keyboard.lua (hunk 2 above)
Comment (weijiuqiao) [reply frontend/ui/data/keyboardlayouts/zh_CN_keyboard.lua]: Made a table that defaults to`{show_candi=true} ` for pinyin settings. As there might be other possible ones in the future (candi separator, use numbers to choose candidates, etc)

**Comment by weijiuqiao** [comment]
>想着以后做一个类似“阅读”的插件，可以配置源的

I'm not sure but maybe there's already something like that in the plugins, that can configure a source.

**Comment by greatyingzi** [comment]
I may need to learn more about this project.

**Comment by Frenzie** [comment]
I don't quite understand what the question is. Something about "cow", the first letter in the alphabet? :-P

**Comment by weijiuqiao** [comment]
>I don't quite understand what the question is. Something about "cow", the first letter in the alphabet? :-P

It's not a question, he was complimenting me for completing this so quickly. The "B" in the slang is a euphemism for a character too vulgar to be written in its original form, which also has a euphemism in Chinese form too :)

**Comment by weijiuqiao** [comment]
Let's merge this and close the issue #9833 ?

**Comment by agugu2000** [comment]
> I don't quite understand what the question is. Something about "cow", the first letter in the alphabet? :-P

see youtube:/watch?v=teQPusK7YFU
you will easily understand this funny slang.

**Comment by agugu2000** [comment]
Many thanks

File: frontend/ui/widget/virtualkeyboard.lua (hunk 1 above)
Comment (agugu2000) [reply frontend/ui/widget/virtualkeyboard.lua:802]: Can we change like this:
 zh = "中文笔画", 
 zh_CN = "简体中文拼音",

Comment (weijiuqiao) [reply frontend/ui/widget/virtualkeyboard.lua:802]: >Can we change like this:
zh = "中文笔画",
zh_CN = "简体中文拼音",

Not really😂 As the file name suggests, they are language names not keyboard names. As to keyboard names, I'm with poire-z on:
>I don't want to complexify that, and have another mapping for keyboard names

**Pull request merged**